[MVC] Remove obsolete APIs

NTaylorMullen opened the pull request:
- Removed ViewAttribute Property
- Removed RazorViewAttribute
- Removed ViewsFeatureProvider
- Removed PageArgumentBinder and its internal implementation DefaultPageArgumentBinder.
- Removed RazorPageAttribute
- Removed corresponding test classes/methods for all the above.
- Reacted to class/member changes in dependencies.

Addresses #7326

rynowak commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Razor/Compilation/CompiledViewDescriptor.cs:
```diff
@@ -37,59 +37,11 @@ public CompiledViewDescriptor(RazorCompiledItem item)
-        /// </summary>
-        /// <param name="item">The <see cref="RazorCompiledItem"/>.</param>
-        /// <param name="attribute">The <see cref="RazorViewAttribute"/>.</param>
-        public CompiledViewDescriptor(RazorCompiledItem item, RazorViewAttribute attribute)
```
This was **just** marked as obsolete. I'm not sure why we'd remove it in 3.0 since it's not costing us anything to keep around.

rynowak commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Razor/Compilation/RazorViewAttribute.cs:
```diff
@@ -1,28 +0,0 @@
-{
-    [Obsolete("This attribute has been superseded by RazorCompiledItem and will not be used by the runtime.")]
-    [AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true)]
-    public class RazorViewAttribute : Attribute
```
This will cause using anything compiled against 2.0 to fail with a `TypeLoadException`. This was also **just** marked obsolete

rynowak replied:
I think we need to take a lot more care with removing obsolete things that appear in user-code and build outputs of user code.

NTaylorMullen replied:
Ah, I took the presence of the release/2.2 link above this specific item verbatim in the issue. Figured it was obsoleted there. Either way I think you have a fair point about the type of complications this would cause to remove an API like this.

rynowak commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Razor/Compilation/ViewsFeatureProvider.cs:
```diff
@@ -1,111 +0,0 @@
-    /// An <see cref="IApplicationFeatureProvider{TFeature}"/> for <see cref="ViewsFeature"/>.
-    /// </summary>
-    [Obsolete("This type is obsolete and will be removed in a future version. See " + nameof(IRazorCompiledItemProvider) + " for alternatives.")]
-    public class ViewsFeatureProvider : IApplicationFeatureProvider<ViewsFeature>
```
For contrast, this is random old extensibility that is a good thing to remove since it will rarely appear in user-code.

NTaylorMullen replied:
One concern I do have, is if I remove this but don't remove the `RazorViewAttribute` and a user doesn't heed the obsolete warnings they'll get a hard to resolve, significant behavior change.

pranavkm replied:
> and a user doesn't heed the obsolete warnings they'll get a hard to resolve

We can add it to the migration notes as something to be aware of. If we get lots of user feedback (I'd think it's fairly unlikely given this attribute comes from Precompilation), we can always change the runtime to throw.

NTaylorMullen replied:
I'm still not convinced since i'd rather break hard or not at all then break soft when it comes to these things (can have migration guidance with either hard or soft breaks). I'll leave it as removed though since you both feel differently 😄 

@danroth27 is there a place where we're collecting migration guidance?

dougbu reviewed (approved):
Once @rynowak's happy, I'm happy

Commit:
[MVC] Remove obsolete APIs

- Removed ViewsFeatureProvider
- Removed PageArgumentBinder and its internal implementation DefaultPageArgumentBinder.
- Removed corresponding test classes/methods for all the above.
- Reacted to class/member changes in dependencies.

#7326

NTaylorMullen commented:
🆙 📅

NTaylorMullen commented:
ping
